source/extensions/config_subscription/grpc/grpc_mux_failover.h

```diff
@@ -0,0 +1,158 @@
+      : currently_used_service_(ServiceOptions::None), grpc_mux_callbacks_(grpc_mux_callbacks),
+        primary_callbacks_(*this),
+        primary_grpc_stream_(std::move(primary_stream_creator(&primary_callbacks_))) {
+    ASSERT(primary_grpc_stream_ != nullptr);
+    // At the moment failover isn't implemented.
+    ASSERT(!failover_stream_creator.has_value());
+  }
+
+  virtual ~GrpcMuxFailover() = default;
+
+  void establishNewStream() {
+    // TODO(adisuissa): At the moment this is a pass-through method. Once the
+    // implementation matures, this call will be updated.
+    primary_grpc_stream_->establishNewStream();
+  }
+
+  bool grpcStreamAvailable() const {
+    // TODO(adisuissa): At the moment this is a pass-through method. Once the
+    // implementation matures, this call will be updated.
+    return primary_grpc_stream_->grpcStreamAvailable();
+  }
+
+  void sendMessage(const RequestType& request) {
+    // TODO(adisuissa): At the moment this is a pass-through method. Once the
+    // implementation matures, this call will be updated.
+    primary_grpc_stream_->sendMessage(request);
+  }
+
+  void maybeUpdateQueueSizeStat(uint64_t size) {
+    // TODO(adisuissa): At the moment this is a pass-through method. Once the
+    // implementation matures, this call will be updated.
+    primary_grpc_stream_->maybeUpdateQueueSizeStat(size);
+  }
+
+  bool checkRateLimitAllowsDrain() {
+    // TODO(adisuissa): At the moment this is a pass-through method. Once the
+    // implementation matures, this call will be updated.
+    return primary_grpc_stream_->checkRateLimitAllowsDrain();
+  }
+
+  absl::optional<Grpc::Status::GrpcStatus> getCloseStatusForTest() {
+    // TODO(adisuissa): At the moment this is a pass-through method. Once the
+    // implementation matures, this call will be updated.
+    return primary_grpc_stream_->getCloseStatusForTest();
+  }
+
+  GrpcStreamInterface<RequestType, ResponseType>& currentStreamForTest() {
+    // TODO(adisuissa): At the moment this is a pass-through method. Once the
+    // implementation matures, this call will be updated.
+    return *primary_grpc_stream_.get();
+  };
+
+private:
+  // The different states the GrpcMuxFailover can be in. Note that the following
+  // are not mutually exclusive, and therefor bitwise value is used. The GrpcMuxFailover
+  // may be connecting to both the primary and failover sources, or connected to
+  // the failover source and connecting to the primary source simultaneously
+  struct ServiceOptions {
+    static constexpr uint32_t None = 0;
+    static constexpr uint32_t ConnectingToPrimary = 0x1;
+    static constexpr uint32_t Primary = 0x2;
+    static constexpr uint32_t ConnectingToFailover = 0x4;
+    static constexpr uint32_t Failover = 0x8;
+  };
+
+  // A helper class that proxies the callbacks of GrpcStreamCallbacks for the primary service.
+  class PrimaryGrpcStreamCallbacksWrapper : public GrpcStreamCallbacks<ResponseType> {
```

Comment:
Why not just `GrpcStreamCallsbacksWrapper`?

Reply:
Changed to `PrimaryGrpcStreamCallbacks` as in a future PR there's going to be `FailoverGrpcStreamCallbacks`.